pocs/camera/camera.py

```diff
@@ -153,6 +174,249 @@ def take_exposure(self, *args, **kwargs):
+                             kwargs={'start_event': coarse_event, 'finished_event': fine_event, **kwargs})
+        fine_thread.start()
+
+        if blocking:
+            fine_event.wait()
+
+        return fine_event
+
+    def _autofocus(self, seconds, focus_range, focus_step, thumbnail_size, coarse, plots,
+                   start_event, finished_event, *args, **kwargs):
+        # If passed a start_event wait until Event is set before proceeding (e.g. wait for coarse focus
+        # to finish before starting fine focus).
+        if start_event:
+            start_event.wait()
+
+        initial_focus = self.focuser.position
+        if coarse:
+            self.logger.debug("Beginning coarse autofocus of {} - initial focus position: {}".format(self,
+                                                                                                     initial_focus))
+        else:
+            self.logger.debug("Beginning autofocus of {} - initial focus position: {}".format(self, initial_focus))
+
+        # Set up paths for temporary focus files, and plots if requested.
+        image_dir = self.config['directories']['images']
+        start_time = current_time(flatten=True)
+        file_path = "{}/{}/{}/{}.{}".format(
+            image_dir,
+            'focus',
+            self.uid,
+            start_time,
+            self.file_extension)
+
+        if plots:
+            # Take an image before focusing, grab a thumbnail from the centre and add it to the plot
+            thumbnail = self._get_thumbnail(seconds, file_path, thumbnail_size)
+            fig = plt.figure(figsize=(9, 18), tight_layout=True)
+            ax1 = fig.add_subplot(3, 1, 1)
+            im1 = ax1.imshow(thumbnail, interpolation='none', cmap='cubehelix')
+            fig.colorbar(im1)
+            ax1.set_title('Initial focus position: {}'.format(initial_focus))
+
+        # Set up encoder positions for autofocus sweep, truncating at focus travel limits if required.
+        if coarse:
+            focus_range = focus_range[1]
+            focus_step = focus_step[1]
+        else:
+            focus_range = focus_range[0]
+            focus_step = focus_step[0]
+
+        focus_positions = np.arange(max(initial_focus - focus_range / 2, self.focuser.min_position),
+                                    min(initial_focus + focus_range / 2, self.focuser.max_position) + 1,
+                                    focus_step, dtype=np.int)
+        n_positions = len(focus_positions)
+
+        f4 = np.empty((n_positions))
+
+        for i, position in enumerate(focus_positions):
+            # Move focus, updating focus_positions with actual encoder position after move.
+            focus_positions[i] = self.focuser.move_to(position)
+
+            # Take exposure
+            thumbnail = self._get_thumbnail(seconds, file_path, thumbnail_size)
+
+            # Calculate Vollath F4 focus metric
+            f4[i] = images.vollath_F4(thumbnail)
```

Comment:
Let's call `images.get_focus_metric(thumbnail, algorithm='vollath_F4')` here. `vollath_F4` can be the default but gives us the ability to add more in the future without too much trouble (although then algorithm should probably be read from the config). `pocs.utils.images` can then just have a some kind of thin-wrapper lookup.